db/seqno_to_time_mapping.cc

```diff
@@ -231,38 +395,69 @@ Status SeqnoToTimeMapping::SeqnoTimePair::Decode(Slice& input) {
   return Status::OK();
 }
 
-bool SeqnoToTimeMapping::Append(SequenceNumber seqno, uint64_t time) {
-  assert(is_sorted_);
+void SeqnoToTimeMapping::CopyFromSeqnoRange(const SeqnoToTimeMapping& src,
+                                            SequenceNumber from_seqno,
+                                            SequenceNumber to_seqno) {
+  bool orig_empty = Empty();
+  auto src_it = src.FindGreaterEqSeqno(from_seqno);
+  // To best answer GetProximalTimeBeforeSeqno(from_seqno) we need an entry
+  // with a seqno before that (if available)
+  if (src_it != src.pairs_.begin()) {
+    --src_it;
+  }
+  auto src_it_end = src.FindGreaterSeqno(to_seqno);
+  std::copy(src_it, src_it_end, std::back_inserter(pairs_));
 
-  // skip seq number 0, which may have special meaning, like zeroed out data
-  if (seqno == 0) {
+  if (!orig_empty || max_time_span_ < UINT64_MAX || capacity_ < UINT64_MAX) {
+    enforced_ = false;
+  }
+}
+
+bool SeqnoToTimeMapping::Append(SequenceNumber seqno, uint64_t time) {
+  if (capacity_ == 0) {
     return false;
   }
-  if (!Empty()) {
-    if (seqno < Last().seqno || time < Last().time) {
-      return false;
-    }
-    if (seqno == Last().seqno) {
-      // Updating Last() would hurt GetProximalSeqnoBeforeTime() queries, so
-      // NOT doing it (for now)
-      return false;
-    }
-    if (time == Last().time) {
-      // Updating Last() here helps GetProximalSeqnoBeforeTime() queries, so
-      // doing it (for now)
-      Last().seqno = seqno;
-      return true;
+  bool added = false;
+  if (seqno == 0) {
+    // skip seq number 0, which may have special meaning, like zeroed out data
+  } else if (pairs_.empty()) {
+    enforced_ = true;
+    pairs_.push_back({seqno, time});
+    // skip normal enforced check below
+    return true;
+  } else {
+    auto& last = pairs_.back();
+    // We can attempt to merge with the last entry if the new entry sorts with
+    // it.
+    if (last.seqno <= seqno) {
+      bool merged = last.Merge({seqno, time});
+      if (!merged) {
+        if (enforced_ && (seqno <= last.seqno || time <= last.time)) {
+          // Out of order append should not happen, except in case of clock
+          // reset
+          assert(false);
+        } else {
+          pairs_.push_back({seqno, time});
+          added = true;
+        }
+      }
+    } else if (!enforced_) {
+      // Treat like AddUnenforced and fix up below
+      pairs_.push_back({seqno, time});
+      added = true;
+    } else {
+      // Out of order append attempted
+      assert(false);
     }
   }
-
-  pairs_.emplace_back(seqno, time);
-
-  if (pairs_.size() > max_capacity_) {
-    // FIXME: be smarter about how we erase to avoid data falling off the
-    // front prematurely.
-    pairs_.pop_front();
+  // Similar to Enforce() but not quite
+  if (!enforced_) {
```

Comment:
very nit: IMHO, this `enforced_` flag are often used to check if the `pairs_` are sorted in many places. And its meaning became a bit ambiguous to me, for example here before `EnforceMaxTimeSpan` and `EnforceCapacity` were reached, `enforced_` is guaranteed to be true. It's a bit confusing why we need to continue to invoke `EnforceSomething` APIs when the class is already `enforced_`. I don't have a concrete suggestion on how to improve this, just want to bring this up to see if I'm not understanding it right.

Reply:
Sorry it is a little confusing.

In the "big step semantics" (between public API operations), enforced_ means that sortedness, timespan and nonstrict capacity invariants are satisfied. Internal to the class during an operation, we might add sorted entries to an object already in the enforced_ state and we know that we only need to fix up the timespan and capacity invariants. So in the "small step semantics" there are some places that only sortedness is guaranteed with enforced_. I think extra state would be excessive, but I can clean this up a bit. Note that `SortAndMerge()` and `EnforceMaxTimeSpan()` and `EnforceCapacity()` are private operations.